### PR TITLE
#12642 Remove 3.9 and promote 3.14 and 3.15

### LINKED ISSUE #12430: ☂️ Add support for Python 3.14
This is the umbrella ticket.

We can use it to discuss various issues related to 3.14

There will be separate issues targeting specific errors

We can close this ticket once we can run the full test suite on Python 3.14 (with all deps) on Linux, Windows and macOS.

Once this is done, we should do a new release

### LINKED ISSUE #12642: Remove GHA for Python 3.9
3.9 is no longer supported... clean up the automated tests...also look into Python 3.14

### PR DESCRIPTION
**Pull request by adiroiban**
## Scope and purpose

Fixes #12642
Fixes #12430

This updates GitHub Actions and prepare for a future release where Python 3.9 is no longer supported.

* It also enabled all Python 3.14 as it looks like all issues were fixed.

* It also enables a few Python 3.15 tests to help adding support for 3.15

* It uses a new release of the test helper... which is waiting a review here https://github.com/twisted/cython-test-exception-raiser/pull/20

* Run tests on Ubuntu 24.04 (from 22.04) by default... I expect that Ubuntu 22.04 will be removed at some time and I 24.04 was required for GTK tests.

* Separate sphinx + pydoctor deps, and don't install them on all dev enviromnents. Most devs don't build the docs... pydoctor apidocs take a very long time to build...and some sphinx extensions are not yet available on Python 3.15

* Enable all Python 3.14 tests

* Enable some Python 3.15 tests ... trial tests are skipped and will be fixed in https://github.com/twisted/twisted/pull/12602

### CODE REVIEW

**Comment by gudnimg** [comment]
Should we also remove these conditions which are now always true? `python_version >= '3.10'"`

https://github.com/twisted/twisted/blob/75facb285e48ead8ab9b6a21310454783d3f14df/pyproject.toml#L113

```
gtk-platform = [
    "twisted[all-non-platform]",
    "pygobject; python_version >= '3.10'",
    "pygobject < 3.52.1; python_version < '3.10'",
]
```

Can be:

```
gtk-platform = [
    "twisted[all-non-platform]",
    "pygobject",
]
```

**Comment by gudnimg** [comment]
Also perhaps update the `[tool.black]` section too

```
 [tool.black]
-target-version = ['py39', 'py310', 'py311', 'py312']
+target-version = ['py310', 'py311', 'py312', 'py313']
```

**Comment by adiroiban** [comment]
@gudnimg thanks

> Should we also remove these conditions which are now always true? python_version >= '3.10'"


Yes. All of that.

Feel free to leave inline comments or push your changes to this branch.

------

This is still WIP ... I am also trying to see where we stand with Python 3.14 and 3.15 ... and pushing various changes between meetings at work

**Comment by adiroiban** [comment]
I will apply the current chagnes mentioned by you... no need to re-add them

**Comment by gudnimg** [comment]
> I will apply the current chagnes mentioned by you... no need to re-add them

If you need any help, I'll be able to do some work over the weekend :) My Linux machine at home has now Ubuntu 26.04 and it ships with Python 3.14. I also have Windows 11 with Python 3.14. No MacOS  though.

**Comment by adiroiban** [comment]
Thanks. A review would help a lot.

Python 3.14 should be ok... all looks good.

I am not checking 3.15...

I see that GitHub has 3.15 beta... but on my `uv` I only have `3.15.0a8`

On my local env the only test that fails is:

trial twisted.trial.test.test_doctest

**Comment by adiroiban** [comment]
@gudnimg  thanks for your feedback.

py313 is not yet supported by the current `black` version that we use.

But I think that it would be nice to move to `ruff` - https://github.com/twisted/twisted/issues/12646

----------

This should be ready for a review

needs-review

File: pyproject.toml
Comment (gudnimg) [inline pyproject.toml]: Shouldn't this be raised to Python 3.10? Otherwise this will allow users to build the package for Python 3.9

**Review by gudnimg — APPROVED** [review]
One minor comment. Besides that the pull request looks good to me. 👍

File: pyproject.toml
Comment (adiroiban) [inline pyproject.toml]: This is the minimum version that we use for testing
```suggestion
requires-python = ">=3.10.20"
```

File: pyproject.toml [hunk 1]
```diff
@@ -17,18 +17,18 @@ dynamic = [
 description = "An asynchronous networking framework written in Python"
 license = { text = "MIT License" }
 # When updating this value, make sure our CI matrix includes a matching minimum version.
-requires-python = ">=3.9.12"
+requires-python = ">=3.10.20"
```
Comment (gudnimg) [inline pyproject.toml]:
```suggestion
requires-python = ">=3.10"
```

----

Hmm I wonder, is Python 3.10.20 not available on Windows?

image

I don't see any executable/installer https://www.python.org/downloads/release/python-31020/

But I see it for Python 3.10.11 https://www.python.org/downloads/release/python-31011/

----

Instead of specifying the minor version, can the toml file just say ">= 3.10"?

Comment (gudnimg) [reply pyproject.toml]: Seems strange to me to put a requirement on the minor version without good reason.

Comment (adiroiban) [reply pyproject.toml]: There was an error. I have fixed it.
In the past, GitHub Python action didn't had all micro-releases...so we were requiring the oldes micro-release available.


The idea was that since we don't test on older micro-releases, it's hard to guarante that it will work.

File: .github/workflows/test.yaml [hunk 2]
```diff
@@ -113,25 +113,21 @@ jobs:
...
             tox-env: 'nodeps-withcov-posix'
             job-name: 'nodeps-test'
-            runs-on: 'ubuntu-22.04'
 
           # `noipv6` is created to make sure all is OK on an OS which doesn't
           # have IPv6 available.
           # Any supported Python version is OK for this job.
           # We go with latest micro release for the smallest minor release
-          # that we support.
+          # that is support by GitHub Actions for the target OS.
```
Comment (glyph) [inline .github/workflows/test.yaml]:
```suggestion
          # that is supported by GitHub Actions for the target OS.
```

File: .github/workflows/test.yaml [hunk 3]
```diff
@@ -113,25 +113,21 @@ jobs:
...
           # This is also used to run tests with minimum dependency versions.
-          - python-version: '3.9'
+          - python-version: '3.10.4'
```
Comment (glyph) [inline .github/workflows/test.yaml]: In principle I guess we should be pinning the micro-version here so we don't get any surprises. But it worries me that we don't have any way to have this upgraded automatically for security fixes, and also why are we going with an old/unsupported version rather than the current (3.10.20)?

File: .github/workflows/test.yaml [hunk 4]
```diff
@@ -305,7 +315,9 @@ jobs:
   coverage-report:
     name: Coverage report
     runs-on: ubuntu-latest
-    if: always()  # Always run the coverage report, even when the tests fail.
+    # Always run the coverage report, even when the tests fail.
+    # As long as the job is not cancelled.
+    if: ${{ !cancelled() }}
```
Comment (glyph) [inline .github/workflows/test.yaml:320]: TIL: https://docs.github.com/en/actions/reference/workflows-and-actions/workflow-cancellation

File: .github/workflows/test.yaml [hunk 5]
```diff
@@ -568,7 +580,7 @@ jobs:
   all-successful:
     # Is very important to force running this always, as otherwise it will be
     #  skipped by default.
-    if: always()
+    if: ${{ !cancelled() }}
```
Comment (glyph) [inline .github/workflows/test.yaml:583]: Just curious, why is the previous version bare but this version is in `{{ }}`?

File: pyproject.toml [hunk 6]
```diff
@@ -110,9 +110,9 @@ all-non-platform = [
 
 macos-platform = [
     "twisted[all-non-platform]",
-    "pyobjc-core >= 12 ; python_version >= '3.10'",
-    "pyobjc-framework-CFNetwork >= 12; python_version >= '3.10'",
-    "pyobjc-framework-Cocoa >= 12; python_version >= '3.10'",
+    "pyobjc-core >= 12",
+    "pyobjc-framework-CFNetwork >= 12",
+    "pyobjc-framework-Cocoa >= 12",
```
Comment (glyph) [inline pyproject.toml:115]: 👍

File: pyproject.toml [hunk 7]
```diff
@@ -279,8 +278,10 @@ include = [
         name = "Misc"
         showcontent = false
 
+# Next time you need to work on black consider migration to ruff
+# https://github.com/twisted/twisted/issues/12646
 [tool.black]
-target-version = ['py39', 'py310', 'py311', 'py312']
+target-version = ['py310', 'py311', 'py312']
```
Comment (glyph) [inline pyproject.toml:284]: Why not expanding upwards to '314'?

**Review by glyph — APPROVED** [review]
Assuming CI is happy, looks good. None of the feedback is blocking.

**Comment by glyph** [comment]
> py313 is not yet supported by the current `black` version that we use.

Oh, I see. Separate PR to upgrade 'black' then. Jeez do we ever need dependabot.

**Comment by adiroiban** [comment]
Thanks for the review.

> Oh, I see. Separate PR to upgrade 'black' then. Jeez do we ever need dependabot.

For `black` a separate PR is better since it might generate formatting changes.

I don't think that we need dependabot... I think that dependabot is useful for end user applications... but for libraries, it's ok to keep support for older dependencies.

now... we might want to keep the dev tools up to date ... so maybe yes. we can enable dependbot to trigger updates for the dev tools, once every 3 months or so

-------

> Just curious, why is the previous version bare but this version is in {{ }}?

Maybe the "bare" version would also work... but I remember there were some issues with the exclamation mark...

File: .github/workflows/test.yaml (hunk 3 above)
Comment (adiroiban) [inline .github/workflows/test.yaml]: I think that we don't need to pin a micro version here... we should good with the latest minor version

```suggestion
          - python-version: '3.10'
```
